projects/coreutils-9.5/man/cp.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.48.5.
.TH CP "1" "March 2024" "GNU coreutils 9.5" "User Commands"
.SH NAME
cp \- copy files and directories
.SH SYNOPSIS
.B cp
[\fI\,OPTION\/\fR]... [\fI\,-T\/\fR] \fI\,SOURCE DEST\/\fR
.br
.B cp
[\fI\,OPTION\/\fR]... \fI\,SOURCE\/\fR... \fI\,DIRECTORY\/\fR
.br
.B cp
[\fI\,OPTION\/\fR]... \fI\,-t DIRECTORY SOURCE\/\fR...
.SH DESCRIPTION
.\" Add any additional description here
.PP
Copy SOURCE to DEST, or multiple SOURCE(s) to DIRECTORY.
.PP
Mandatory arguments to long options are mandatory for short options too.
.TP
\fB\-a\fR, \fB\-\-archive\fR
same as \fB\-dR\fR \fB\-\-preserve\fR=\fI\,all\/\fR
.TP
\fB\-\-attributes\-only\fR
don't copy the file data, just the attributes
.TP
\fB\-\-backup\fR[=\fI\,CONTROL\/\fR]
make a backup of each existing destination file
.TP
\fB\-b\fR
like \fB\-\-backup\fR but does not accept an argument
.TP
\fB\-\-copy\-contents\fR
copy contents of special files when recursive
.TP
\fB\-d\fR
same as \fB\-\-no\-dereference\fR \fB\-\-preserve\fR=\fI\,links\/\fR
.TP
\fB\-\-debug\fR
explain how a file is copied.  Implies \fB\-v\fR
.TP
\fB\-f\fR, \fB\-\-force\fR
if an existing destination file cannot be
opened, remove it and try again (this option
is ignored when the \fB\-n\fR option is also used)
.TP
\fB\-i\fR, \fB\-\-interactive\fR
prompt before overwrite (overrides a previous \fB\-n\fR
option)
.TP
\fB\-H\fR
follow command\-line symbolic links in SOURCE
.TP
\fB\-l\fR, \fB\-\-link\fR
hard link files instead of copying
.TP
\fB\-L\fR, \fB\-\-dereference\fR
always follow symbolic links in SOURCE
.TP
\fB\-n\fR, \fB\-\-no\-clobber\fR
(deprecated) silently skip existing files.
See also \fB\-\-update\fR
.TP
\fB\-P\fR, \fB\-\-no\-dereference\fR
never follow symbolic links in SOURCE
.TP
\fB\-p\fR
same as \fB\-\-preserve\fR=\fI\,mode\/\fR,ownership,timestamps
.TP
\fB\-\-preserve\fR[=\fI\,ATTR_LIST\/\fR]
preserve the specified attributes
.TP
\fB\-\-no\-preserve\fR=\fI\,ATTR_LIST\/\fR
don't preserve the specified attributes
.TP
\fB\-\-parents\fR
use full source file name under DIRECTORY
.TP
\fB\-R\fR, \fB\-r\fR, \fB\-\-recursive\fR
copy directories recursively
.TP
\fB\-\-reflink\fR[=\fI\,WHEN\/\fR]
control clone/CoW copies. See below
.TP
\fB\-\-remove\-destination\fR
remove each existing destination file before
attempting to open it (contrast with \fB\-\-force\fR)
.TP
\fB\-\-sparse\fR=\fI\,WHEN\/\fR
control creation of sparse files. See below
.TP
\fB\-\-strip\-trailing\-slashes\fR
remove any trailing slashes from each SOURCE
argument
.TP
\fB\-s\fR, \fB\-\-symbolic\-link\fR
make symbolic links instead of copying
.TP
\fB\-S\fR, \fB\-\-suffix\fR=\fI\,SUFFIX\/\fR
override the usual backup suffix
.TP
\fB\-t\fR, \fB\-\-target\-directory\fR=\fI\,DIRECTORY\/\fR
copy all SOURCE arguments into DIRECTORY
.TP
\fB\-T\fR, \fB\-\-no\-target\-directory\fR
treat DEST as a normal file
.TP
\fB\-\-update\fR[=\fI\,UPDATE\/\fR]
control which existing files are updated;
UPDATE={all,none,none\-fail,older(default)}.
.TP
\fB\-u\fR
equivalent to \fB\-\-update\fR[=\fI\,older\/\fR].  See below
.TP
\fB\-v\fR, \fB\-\-verbose\fR
explain what is being done
.TP
\fB\-\-keep\-directory\-symlink\fR
follow existing symlinks to directories
.TP
\fB\-x\fR, \fB\-\-one\-file\-system\fR
stay on this file system
.TP
\fB\-Z\fR
set SELinux security context of destination
file to default type
.TP
\fB\-\-context\fR[=\fI\,CTX\/\fR]
like \fB\-Z\fR, or if CTX is specified then set the
SELinux or SMACK security context to CTX
.TP
\fB\-\-help\fR
display this help and exit
.TP
\fB\-\-version\fR
output version information and exit
.PP
ATTR_LIST is a comma\-separated list of attributes. Attributes are 'mode' for
permissions (including any ACL and xattr permissions), 'ownership' for user
and group, 'timestamps' for file timestamps, 'links' for hard links, 'context'
for security context, 'xattr' for extended attributes, and 'all' for all
attributes.
.PP
By default, sparse SOURCE files are detected by a crude heuristic and the
corresponding DEST file is made sparse as well.  That is the behavior
selected by \fB\-\-sparse\fR=\fI\,auto\/\fR.  Specify \fB\-\-sparse\fR=\fI\,always\/\fR to create a sparse DEST
file whenever the SOURCE file contains a long enough sequence of zero bytes.
Use \fB\-\-sparse\fR=\fI\,never\/\fR to inhibit creation of sparse files.
.PP
UPDATE controls which existing files in the destination are replaced.
\&'all' is the default operation when an \fB\-\-update\fR option is not specified,
and results in all existing files in the destination being replaced.
\&'none' is like the \fB\-\-no\-clobber\fR option, in that no files in the
destination are replaced, and skipped files do not induce a failure.
\&'none\-fail' also ensures no files are replaced in the destination,
but any skipped files are diagnosed and induce a failure.
\&'older' is the default operation when \fB\-\-update\fR is specified, and results
in files being replaced if they're older than the corresponding source file.
.PP
When \fB\-\-reflink\fR[=\fI\,always\/\fR] is specified, perform a lightweight copy, where the
data blocks are copied only when modified.  If this is not possible the copy
fails, or if \fB\-\-reflink\fR=\fI\,auto\/\fR is specified, fall back to a standard copy.
Use \fB\-\-reflink\fR=\fI\,never\/\fR to ensure a standard copy is performed.
.PP
The backup suffix is '~', unless set with \fB\-\-suffix\fR or SIMPLE_BACKUP_SUFFIX.
The version control method may be selected via the \fB\-\-backup\fR option or through
the VERSION_CONTROL environment variable.  Here are the values:
.TP
none, off
never make backups (even if \fB\-\-backup\fR is given)
.TP
numbered, t
make numbered backups
.TP
existing, nil
numbered if numbered backups exist, simple otherwise
.TP
simple, never
always make simple backups
.PP
As a special case, cp makes a backup of SOURCE when the force and backup
options are given and SOURCE and DEST are the same name for an existing,
regular file.
.SH AUTHOR
Written by Torbjorn Granlund, David MacKenzie, and Jim Meyering.
.SH "REPORTING BUGS"
GNU coreutils online help: <https://www.gnu.org/software/coreutils/>
.br
Report any translation bugs to <https://translationproject.org/team/>
.SH COPYRIGHT
Copyright \(co 2024 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
\fBinstall\fP(1)
.PP
.br
Full documentation <https://www.gnu.org/software/coreutils/cp>
.br
or available locally via: info \(aq(coreutils) cp invocation\(aq


projects/busybox-1.37.0/coreutils/cp.c
/* vi: set sw=4 ts=4: */
/*
 * Mini cp implementation for busybox
 *
 * Copyright (C) 2000 by Matt Kraai <kraai@alumni.carnegiemellon.edu>
 * SELinux support by Yuichi Nakamura <ynakam@hitachisoft.jp>
 *
 * Licensed under GPLv2 or later, see file LICENSE in this source tree.
 */
/* Mar 16, 2003      Manuel Novoa III   (mjn3@codepoet.org)
 *
 * Size reduction.
 */
//config:config CP
//config:	bool "cp (10 kb)"
//config:	default y
//config:	help
//config:	cp is used to copy files and directories.
//config:
//config:config FEATURE_CP_LONG_OPTIONS
//config:	bool "Enable long options"
//config:	default y
//config:	depends on CP && LONG_OPTS
//config:	help
//config:	Enable long options.
//config:	Also add support for --parents option.
//config:
//config:config FEATURE_CP_REFLINK
//config:	bool "Enable --reflink[=auto]"
//config:	default y
//config:	depends on FEATURE_CP_LONG_OPTIONS

//applet:IF_CP(APPLET_NOEXEC(cp, cp, BB_DIR_BIN, BB_SUID_DROP, cp))
/* NOEXEC despite cases when it can be a "runner" (cp -r LARGE_DIR NEW_DIR) */

//kbuild:lib-$(CONFIG_CP) += cp.o

/* http://www.opengroup.org/onlinepubs/007904975/utilities/cp.html */

// Options of cp from GNU coreutils 6.10:
// -a, --archive
// -f, --force
// -i, --interactive
// -l, --link
// -L, --dereference
// -P, --no-dereference
// -R, -r, --recursive
// -s, --symbolic-link
// -v, --verbose
// -H	follow command-line symbolic links in SOURCE
// -d	same as --no-dereference --preserve=links
// -p	same as --preserve=mode,ownership,timestamps
// -c	same as --preserve=context
// -u, --update
//	copy only when the SOURCE file is newer than the destination
//	file or when the destination file is missing
// --remove-destination
//	remove each existing destination file before attempting to open
// --parents
//	use full source file name under DIRECTORY
// -T, --no-target-directory
//	treat DEST as a normal file
// NOT SUPPORTED IN BBOX:
// --backup[=CONTROL]
//	make a backup of each existing destination file
// -b	like --backup but does not accept an argument
// --copy-contents
//	copy contents of special files when recursive
// --preserve[=ATTR_LIST]
//	preserve attributes (default: mode,ownership,timestamps),
//	if possible additional attributes: security context,links,all
// --no-preserve=ATTR_LIST
// --sparse=WHEN
//	control creation of sparse files
// --strip-trailing-slashes
//	remove any trailing slashes from each SOURCE argument
// -S, --suffix=SUFFIX
//	override the usual backup suffix
// -t, --target-directory=DIRECTORY
//	copy all SOURCE arguments into DIRECTORY
// -x, --one-file-system
//	stay on this file system
// -Z, --context=CONTEXT
//	(SELinux) set SELinux security context of copy to CONTEXT

//usage:#define cp_trivial_usage
//usage:       "[-arPLHpfinlsTu] SOURCE DEST\n"
//usage:       "or: cp [-arPLHpfinlsu] SOURCE... { -t DIRECTORY | DIRECTORY }"
//usage:#define cp_full_usage "\n\n"
//usage:       "Copy SOURCEs to DEST\n"
//usage:     "\n	-a	Same as -dpR"
//usage:	IF_SELINUX(
//usage:     "\n	-c	Preserve security context"
//usage:	)
//usage:     "\n	-R,-r	Recurse"
//usage:     "\n	-d,-P	Preserve symlinks (default if -R)"
//usage:     "\n	-L	Follow all symlinks"
//usage:     "\n	-H	Follow symlinks on command line"
//usage:     "\n	-p	Preserve file attributes if possible"
//usage:     "\n	-f	Overwrite"
//usage:     "\n	-i	Prompt before overwrite"
//usage:     "\n	-n	Don't overwrite"
//usage:     "\n	-l,-s	Create (sym)links"
//usage:     "\n	-T	Refuse to copy if DEST is a directory"
//usage:     "\n	-t DIR	Copy all SOURCEs into DIR"
//usage:     "\n	-u	Copy only newer files"

#include "libbb.h"
#include "libcoreutils/coreutils.h"

/* This is a NOEXEC applet. Be very careful! */

int cp_main(int argc, char **argv) MAIN_EXTERNALLY_VISIBLE;
int cp_main(int argc, char **argv)
{
	struct stat source_stat;
	struct stat dest_stat;
	const char *last;
	const char *dest;
	int s_flags;
	int d_flags;
	int flags;
	int status;
#if ENABLE_FEATURE_CP_LONG_OPTIONS
	enum {
		/*OPT_rmdest  = FILEUTILS_RMDEST = 1 << FILEUTILS_CP_OPTBITS */
		OPT_parents = 1 << (FILEUTILS_CP_OPTBITS+1),
		OPT_reflink = 1 << (FILEUTILS_CP_OPTBITS+2),
	};
# if ENABLE_FEATURE_CP_REFLINK
	char *reflink = NULL;
# endif
	flags = getopt32long(argv, "^"
		FILEUTILS_CP_OPTSTR
		"\0"
		// Need at least one argument. (Usually two+, but -t DIR can have only one)
		// Soft- and hardlinking doesn't mix
		// -P and -d are the same (-P is POSIX, -d is GNU)
		// -r and -R are the same
		// -R (and therefore -r) turns on -d (coreutils does this)
		// -a = -pdR
		// -i overrides -n and vice versa (last wins)
		"-1:l--s:s--l:Pd:rRd:Rd:apdR:i-n:n-i",
		"archive\0"        No_argument "a"
		"force\0"          No_argument "f"
		"interactive\0"    No_argument "i"
		"no-clobber\0"     No_argument "n"
		"link\0"           No_argument "l"
		"dereference\0"    No_argument "L"
		"no-dereference\0" No_argument "P"
		"recursive\0"      No_argument "R"
		"symbolic-link\0"  No_argument "s"
		"no-target-directory\0" No_argument "T"
		"target-directory\0" Required_argument "t"
		"verbose\0"        No_argument "v"
		"update\0"         No_argument "u"
		"remove-destination\0" No_argument "\xff"
		"parents\0"        No_argument "\xfe"
# if ENABLE_FEATURE_CP_REFLINK
		"reflink\0"        Optional_argument "\xfd"
# endif
		, &last
# if ENABLE_FEATURE_CP_REFLINK
		, &reflink
# endif
	);
# if ENABLE_FEATURE_CP_REFLINK
	BUILD_BUG_ON((int)OPT_reflink != (int)FILEUTILS_REFLINK);
	if (flags & FILEUTILS_REFLINK) {
		if (!reflink)
			flags |= FILEUTILS_REFLINK_ALWAYS;
		else if (strcmp(reflink, "always") == 0)
			flags |= FILEUTILS_REFLINK_ALWAYS;
		else if (strcmp(reflink, "auto") != 0)
			bb_show_usage();
	}
# endif
#else
	flags = getopt32(argv, "^"
		FILEUTILS_CP_OPTSTR
		"\0"
		"-1:l--s:s--l:Pd:rRd:Rd:apdR"
		, &last
	);
#endif
	argc -= optind;
	argv += optind;
	/* Reverse this bit. If there is -d, bit is not set: */
	flags ^= FILEUTILS_DEREFERENCE;
	/* coreutils 6.9 compat:
	 * by default, "cp" derefs symlinks (creates regular dest files),
	 * but "cp -R" does not. We switch off deref if -r or -R (see above).
	 * However, "cp -RL" must still deref symlinks: */
	if (flags & FILEUTILS_DEREF_SOFTLINK) /* -L */
		flags |= FILEUTILS_DEREFERENCE;

#if ENABLE_SELINUX
	if (flags & FILEUTILS_PRESERVE_SECURITY_CONTEXT) {
		selinux_or_die();
	}
#endif

	status = EXIT_SUCCESS;
	if (!(flags & FILEUTILS_TARGET_DIR)) {
		last = argv[argc - 1];
		if (argc < 2)
			bb_show_usage();
		if (argc != 2) {
			if (flags & FILEUTILS_NO_TARGET_DIR)
				bb_show_usage();
			/* "cp A B C... DIR" - target must be dir */
		} else /* argc == 2 */ {
			/* "cp A B" - only case where target can be not a dir */
			s_flags = cp_mv_stat2(*argv, &source_stat,
					(flags & FILEUTILS_DEREFERENCE) ? stat : lstat);
			if (s_flags < 0) /* error other than ENOENT */
				return EXIT_FAILURE;
			d_flags = cp_mv_stat(last, &dest_stat);
			if (d_flags < 0) /* error other than ENOENT */
				return EXIT_FAILURE;

			if (flags & FILEUTILS_NO_TARGET_DIR) { /* -T */
				if (!(s_flags & 2) && (d_flags & 2))
					/* cp -T NOTDIR DIR */
					bb_error_msg_and_die("'%s' is a directory", last);
			}

#if ENABLE_FEATURE_CP_LONG_OPTIONS
			//bb_error_msg("flags:%x FILEUTILS_RMDEST:%x OPT_parents:%x",
			//	flags, FILEUTILS_RMDEST, OPT_parents);
			if (flags & OPT_parents) {
				if (!(d_flags & 2)) {
					bb_simple_error_msg_and_die("with --parents, the destination must be a directory");
				}
			}
			if (flags & FILEUTILS_RMDEST) {
				flags |= FILEUTILS_FORCE;
			}
#endif

			/* ...if neither is a directory...  */
			if (!((s_flags | d_flags) & 2)
			    /* ...or: recursing, the 1st is a directory, and the 2nd doesn't exist... */
			 || ((flags & FILEUTILS_RECUR) && (s_flags & 2) && !d_flags)
			 || (flags & FILEUTILS_NO_TARGET_DIR)
			) {
				/* Do a simple copy */
				dest = last;
				goto DO_COPY; /* NB: argc==2 -> *++argv==last */
			}
		}
	}
	/* else: last is DIR from "-t DIR" */

	while (1) {
#if ENABLE_FEATURE_CP_LONG_OPTIONS
		if (flags & OPT_parents) {
			char *dest_dup;
			char *dest_dir;
			dest = concat_path_file(last, *argv);
			dest_dup = xstrdup(dest);
			dest_dir = dirname(dest_dup);
			if (bb_make_directory(dest_dir, -1, FILEUTILS_RECUR)) {
				return EXIT_FAILURE;
			}
			free(dest_dup);
			goto DO_COPY;
		}
#endif
		dest = concat_path_file(last, bb_get_last_path_component_strip(*argv));
 DO_COPY:
		if (copy_file(*argv, dest, flags) < 0) {
			status = EXIT_FAILURE;
		}
		if (!*++argv || *argv == last) {
			/* possibly leaking dest... */
			break;
		}
		/* don't move up: dest may be == last and not malloced! */
		free((void*)dest);
	}

	/* Exit. We are NOEXEC, not NOFORK. We do exit at the end of main() */
	return status;
}
